src/main/flight/autopilot.c

```diff
@@ -22,30 +22,94 @@
 
 #include "platform.h"
 #include "build/debug.h"
+#include "common/filter.h"
 #include "common/maths.h"
+#include "fc/rc.h"
+#include "fc/runtime_config.h"
 
 #include "flight/imu.h"
-#include "flight/pid.h"
 #include "flight/position.h"
 #include "rx/rx.h"
+#include "sensors/gyro.h"
+#include "sensors/compass.h"
 
 #include "autopilot.h"
 
 #define ALTITUDE_P_SCALE  0.01f
 #define ALTITUDE_I_SCALE  0.003f
 #define ALTITUDE_D_SCALE  0.01f
 #define ALTITUDE_F_SCALE  0.01f
+#define POSITION_P_SCALE  0.001f
+#define POSITION_I_SCALE  0.0004f
+#define POSITION_D_SCALE  0.003f
+#define POSITION_J_SCALE  0.0008f
 
 static pidCoefficient_t altitudePidCoeffs;
+static pidCoefficient_t positionPidCoeffs;
+
 static float altitudeI = 0.0f;
 static float throttleOut = 0.0f;
 
+typedef struct {
+    uint32_t distanceCm;
+    uint32_t previousDistanceCm;
+    float previousDistancePitch;
+    float previousDistanceRoll;
+    float previousVelocityPitch;
+    float previousVelocityRoll;
+    float previousHeading;
+    float pitchI;
+    float rollI;
+    bool justStarted;
+} posHoldState;
+
+static posHoldState posHold = {
+    .distanceCm = 0,
+    .previousDistanceCm = 0,
+    .previousDistancePitch = 0.0f,
+    .previousDistanceRoll = 0.0f,
+    .previousVelocityPitch = 0.0f,
+    .previousVelocityRoll = 0.0f,
+    .previousHeading = 0.0f,
+    .pitchI = 0.0f,
+    .rollI = 0.0f,
+    .justStarted = true,
+};
+
+static gpsLocation_t currentTargetLocation = {0, 0, 0};
+float posHoldAngle[ANGLE_INDEX_COUNT];
+static pt1Filter_t velocityPitchLpf;
+static pt1Filter_t accelerationPitchLpf;
+static pt1Filter_t velocityRollLpf;
+static pt1Filter_t accelerationRollLpf;
+static float positionLpfCutoffHz;
+
 void autopilotInit(const autopilotConfig_t *config)
 {
     altitudePidCoeffs.Kp = config->altitude_P * ALTITUDE_P_SCALE;
     altitudePidCoeffs.Ki = config->altitude_I * ALTITUDE_I_SCALE;
     altitudePidCoeffs.Kd = config->altitude_D * ALTITUDE_D_SCALE;
     altitudePidCoeffs.Kf = config->altitude_F * ALTITUDE_F_SCALE;
+    positionPidCoeffs.Kp = config->position_P * POSITION_P_SCALE;
+    positionPidCoeffs.Ki = config->position_I * POSITION_I_SCALE;
+    positionPidCoeffs.Kd = config->position_D * POSITION_D_SCALE;
+    positionPidCoeffs.Kf = config->position_J * POSITION_J_SCALE; // Kf used for acceleration
+    positionLpfCutoffHz = config->position_cutoff / 100.0f;
+    const float gain = pt1FilterGain(positionLpfCutoffHz, 0.1f); // assume 10Hz GPS connection at start
+    pt1FilterInit(&velocityPitchLpf, gain);
+    pt1FilterInit(&accelerationPitchLpf, gain);
+    pt1FilterInit(&velocityRollLpf, gain);
+    pt1FilterInit(&accelerationRollLpf, gain);
+}
+
+const pidCoefficient_t *getAltitudePidCoeffs(void)
+{
+    return &altitudePidCoeffs;
+}
+
+const pidCoefficient_t *getPositionPidCoeffs(void)
+{
+    return &positionPidCoeffs;
```

Comment:
I think you forgot to remove these functions (they are not part of the header file)

Reply:
Yes, sorry.